### PR TITLE
Feature/file monitoring

### PR DESCRIPTION
**Pull request by ChristianMorup**
Initial merge of the file monitoring feature. 

Hopefully this will enable the possibility to start working on some of the other tasks.

### CODE REVIEW

File: RTDS/RTDS.IntegrationTest/Monitoring/Wrapper/Test_FileSystemWatcherWrapper.cs [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+            DirectoryInfo testDirectory = new DirectoryInfo(_testDataPath);
+
+            foreach (FileInfo file in testDirectory.GetFiles())
+            {
+                file.Delete();
+            }
+            foreach (DirectoryInfo directory in testDirectory.GetDirectories())
+            {
+                directory.Delete(true);
+            }
```
Comment (Cpilgaard) [inline RTDS/RTDS.IntegrationTest/Monitoring/Wrapper/Test_FileSystemWatcherWrapper.cs]: Bare af interesse: Slettes filerne ikke automatisk, hvis directory slettes? :-)

File: RTDS/RTDS.IntegrationTest/Monitoring/Wrapper/Test_FileSystemWatcherWrapper.cs [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+            Assert.That(_watcher.EnableRaisingEvents, Is.True);
+            Assert.That(_uut.EnableRaisingEvents, Is.True);
+        }
+
+        [Test]
+        public void Dispose_DisposesTimer_TimerIsDisposed()
+        {
+            //Arrange: 
+            bool isDisposed = false;
+            _watcher.Disposed += (sender, args) => { isDisposed = true; };
```
Comment (Cpilgaard) [inline RTDS/RTDS.IntegrationTest/Monitoring/Wrapper/Test_FileSystemWatcherWrapper.cs]: Er ikke helt familiar med lambda udtrykkene

File: RTDS/RTDS.IntegrationTest/Monitoring/Wrapper/Test_FileSystemWatcherWrapper.cs [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+            //Assert:
+            Assert.That(isDisposed, Is.True);
+        }
+
+        private void OnCreated(object source, FileSystemEventArgs e)
+        {
+            _eventReceived = true;
+        }
+
```
Comment (Cpilgaard) [inline RTDS/RTDS.IntegrationTest/Monitoring/Wrapper/Test_FileSystemWatcherWrapper.cs]: Minus Spacing

File: RTDS/RTDS.IntegrationTest/Monitoring/Wrapper/Test_FileSystemWatcherWrapper.cs [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+            Assert.That(_watcher.EnableRaisingEvents, Is.True);
+            Assert.That(_uut.EnableRaisingEvents, Is.True);
+        }
+
+        [Test]
+        public void Dispose_DisposesTimer_TimerIsDisposed()
+        {
+            //Arrange: 
+            bool isDisposed = false;
```
Comment (Cpilgaard) [inline RTDS/RTDS.IntegrationTest/Monitoring/Wrapper/Test_FileSystemWatcherWrapper.cs]: Thought we were going to use "var" more often? ;-)

File: RTDS/RTDS.UnitTest/Monitoring/Test_FileHandler.cs [hunk 5]
```diff
@@ -0,0 +1,7 @@
+﻿namespace RTDS.UnitTest.Monitoring
+{
+    class Test_FileHandler
+    {
+
+    }
+}
```
Comment (Cpilgaard) [inline RTDS/RTDS.UnitTest/Monitoring/Test_FileHandler.cs]: Empty Class?

File: RTDS/RTDS.UnitTest/Monitoring/Test_FileMonitor.cs [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+
+        [Test]
+        public void StartMonitoringAsync_NullArgumentIsPassed_ThrowsException()
+        {
+            //Act + Assert:
+            Assert.Throws<ArgumentNullException>(() => _uut.StartMonitoringAsync(null));
+        }
+
+        [TestCase("ThisIsAValidPath")]
+        [TestCase("ThisIsAlsoAValidPath")]
```
Comment (Cpilgaard) [inline RTDS/RTDS.UnitTest/Monitoring/Test_FileMonitor.cs]: Why two?

File: RTDS/RTDS.IntegrationTest/Monitoring/Wrapper/Test_FileSystemWatcherWrapper.cs [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+        {
+            DirectoryInfo testDirectory = new DirectoryInfo(_testDataPath);
+
+            foreach (FileInfo file in testDirectory.GetFiles())
+            {
+                file.Delete();
+            }
+            foreach (DirectoryInfo directory in testDirectory.GetDirectories())
+            {
+                directory.Delete(true);
```
[the remark this answers is not in the document]
Comment (ChristianMorup) [reply RTDS/RTDS.IntegrationTest/Monitoring/Wrapper/Test_FileSystemWatcherWrapper.cs]: That is possible. I'm not sure. For now it is atleast quite explicit.

File: RTDS/RTDS.IntegrationTest/Monitoring/Wrapper/Test_FileSystemWatcherWrapper.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (ChristianMorup) [reply RTDS/RTDS.IntegrationTest/Monitoring/Wrapper/Test_FileSystemWatcherWrapper.cs]: It doesn't make any difference. It is made a bool at compile time.

File: RTDS/RTDS.UnitTest/Monitoring/Test_FileMonitor.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (ChristianMorup) [reply RTDS/RTDS.UnitTest/Monitoring/Test_FileMonitor.cs]: Because I could make a test that just pass an exact string as a valid path.

File: RTDS/RTDS/Monitoring/Args/FileMonitorFinishedArgs.cs [hunk 8]
```diff
@@ -0,0 +1,12 @@
+﻿namespace RTDS.Monitoring.Args
+{
+    internal class FileMonitorFinishedArgs
+    {
+        public FileMonitorFinishedArgs(IFileMonitor monitor)
```
[the remark this answers is not in the document]
Comment (ChristianMorup) [reply RTDS/RTDS/Monitoring/Args/FileMonitorFinishedArgs.cs]: We need to wrap the file monitor, so we can cancel and dispose it when it is finished. This is done by wrapping it in the args.

File: RTDS/RTDS/Monitoring/Args/SearchDirectoryArgs.cs [hunk 9]
```diff
@@ -0,0 +1,17 @@
+﻿using System;
+
+namespace RTDS.Monitoring.Args
+{
+    internal class SearchDirectoryArgs : EventArgs
+    {
+        public SearchDirectoryArgs(string path, string name)
```
[the remark this answers is not in the document]
Comment (ChristianMorup) [reply RTDS/RTDS/Monitoring/Args/SearchDirectoryArgs.cs]: I'm just wrapping the path and the name of the folder that has been created so that I know what folder that I should start monitoring.

File: RTDS/RTDS/Monitoring/FileMonitor.cs [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+        private readonly IFileSystemWatcherWrapper _watcher;
+        private readonly ITimerWrapper _timer;
+
+        public FileMonitor(IFileSystemWatcherWrapper watcher, ITimerWrapper timer)
+        {
+            _watcher = watcher;
+            _timer = timer;
+        }
+
+        public Task StartMonitoringAsync(string path)
```
[the remark this answers is not in the document]
Comment (ChristianMorup) [reply RTDS/RTDS/Monitoring/FileMonitor.cs]: It is a naming convention that async public methods are suffixed with async. See https://docs.microsoft.com/en-us/dotnet/standard/asynchronous-programming-patterns/task-based-asynchronous-pattern-tap?redirectedfrom=MSDN .

File: RTDS/RTDS.UnitTest/Monitoring/Test_FolderMonitor.cs [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+
+        [Test]
+        public void StartMonitoringAsync_NullArgumentIsPassed_ThrowsException()
+        {
+            //Act + Assert:
+            Assert.Throws<ArgumentNullException>(() => _uut.StartMonitoringAsync(null));
+        }
+
+        [TestCase("ThisIsAValidPath")]
+        [TestCase("ThisIsAlsoAValidPath")]
```
[the remark this answers is not in the document]
Comment (ChristianMorup) [reply RTDS/RTDS.UnitTest/Monitoring/Test_FolderMonitor.cs]: Because I could make a test that just pass an exact string as a valid path.

File: RTDS/RTDS.IntegrationTest/Monitoring/Wrapper/Test_FileSystemWatcherWrapper.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (ChristianMorup) [reply RTDS/RTDS.IntegrationTest/Monitoring/Wrapper/Test_FileSystemWatcherWrapper.cs]: += is just an event subscription. You can read about the lamda part here: https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/statements-expressions-operators/lambda-expressions

File: RTDS/RTDS/Monitoring/IFileSystemWatcherWrapper.cs [hunk 12]
```diff
@@ -0,0 +1,18 @@
+﻿using System;
+using System.IO;
+
+namespace RTDS.Monitoring
+{
+    public interface IFileSystemWatcherWrapper : IDisposable
+    {
+        string Path { get; set; }
+
+        event FileSystemEventHandler Changed;
```
Comment (MikkelHarbo) [inline RTDS/RTDS/Monitoring/IFileSystemWatcherWrapper.cs]: Is this the method that handles the respons of FileSystemWatcher, or what is it used for?

File: RTDS/RTDS.Application/Program.cs [hunk 13]
```diff
@@ -7,16 +7,18 @@ namespace RTDS.Application
...
-            IFileSystemWatcherWrapper watcher = new FileSystemWatcherWrapper(new FileSystemWatcher());
-            FolderMonitor folderMonitor = new FolderMonitor(watcher);
+            var systemwatcher = new FileSystemWatcher();
+            IFileSystemWatcherWrapper watcher = new FileSystemWatcherWrapper(systemwatcher);
+            folderMonitor = new FolderMonitor(watcher);
 
             folderMonitor.FolderCreated += OnFolderCreated;
             
-            folderMonitor.StarMonitoringAsync("C:\\Users\\herni\\OneDrive\\Skrivebord\\RTDS_Experimental_Test");
+            folderMonitor.StarMonitoringAsync("C:\\Users\\chrmo\\Desktop\\RTDS");
```
Comment (MikkelHarbo) [inline RTDS/RTDS.Application/Program.cs]: Does this mean, that the StartMonitoringAsync, is monitoring a folder, for new folders or files?

File: RTDS/RTDS/Monitoring/FileSystemWatcherWrapper.cs [hunk 14]
```diff
@@ -7,19 +7,21 @@
...
-    public class FileSystemWatcherWrapper : IFileSystemWatcherWrapper
+    internal class FileSystemWatcherWrapper : IFileSystemWatcherWrapper
     {
         private readonly FileSystemWatcher _watcher;
         public event FileSystemEventHandler Created;
+
         public string Path
         {
             get => _watcher.Path;
             set => _watcher.Path = value;
```
Comment (MikkelHarbo) [inline RTDS/RTDS/Monitoring/FileSystemWatcherWrapper.cs]: What is the "Value", and which function does it have?

File: RTDS/RTDS/Monitoring/FileSystemWatcherWrapper.cs [hunk 15]
```diff
@@ -7,19 +7,21 @@
...
         }
 
         public string Filter { get; set; }
-        public NotifyFilters NotifyFilters 
-        { 
+
+        public NotifyFilters NotifyFilters
+        {
             get => _watcher.NotifyFilter;
             set => _watcher.NotifyFilter = value;
```
Comment (MikkelHarbo) [inline RTDS/RTDS/Monitoring/FileSystemWatcherWrapper.cs]: Same question as above.

File: RTDS/RTDS/Monitoring/FolderMonitor.cs [hunk 16]
```diff
@@ -6,7 +6,7 @@ namespace RTDS.Monitoring
 {
     public class FolderMonitor
     {
-        private IFileSystemWatcherWrapper _systemWatcher;
+        private readonly IFileSystemWatcherWrapper _systemWatcher;
```
Comment (MikkelHarbo) [inline RTDS/RTDS/Monitoring/FolderMonitor.cs]: What difference does it make, whether this is readonly or not? It is private?

File: RTDS/RTDS.Application/Program.cs [hunk 17]
```diff
@@ -2,24 +2,23 @@
...
-        private static FolderMonitor folderMonitor;
+        private static MonitorController _monitorController;
         static void Main(string[] args)
         {
-            var systemwatcher = new FileSystemWatcher();
-            IFileSystemWatcherWrapper watcher = new FileSystemWatcherWrapper(systemwatcher);
-            folderMonitor = new FolderMonitor(watcher);
+            IFileSystemWatcherWrapper watcher = new FileSystemWatcherWrapper(new FileSystemWatcher());
+            IMonitor folderMonitor = new FolderMonitor(watcher);
+            _monitorController = new MonitorController(folderMonitor, new MonitorFactory());
```
Comment (MikkelHarbo) [inline RTDS/RTDS.Application/Program.cs]: Nice refactoring.

File: RTDS/RTDS.IntegrationTest/Monitoring/Test_FileSystemWatcherWrapper.cs [hunk 18]
```diff
@@ -0,0 +1,80 @@
...
+
+            foreach (FileInfo file in testDirectory.GetFiles())
+            {
+                file.Delete();
+            }
+            foreach (DirectoryInfo directory in testDirectory.GetDirectories())
+            {
+                directory.Delete(true);
+            }
+        }
```
Comment (MikkelHarbo) [inline RTDS/RTDS.IntegrationTest/Monitoring/Test_FileSystemWatcherWrapper.cs]: Does this work like a "Clear-up" after each test?

File: RTDS/RTDS.UnitTest/Monitoring/Test_FileMonitor.cs [hunk 19]
```diff
@@ -0,0 +1,115 @@
...
+        }
+
+        [Test]
+        public void StartMonitoringAsync_AddingHandlers_CorrectHandlersAreAdded()
+        {
+            //Act:
+            Task task = _uut.StartMonitoringAsync("ValidPath");
+
+            //Assert:
+            AssertCompletion(task);
```
Comment (MikkelHarbo) [inline RTDS/RTDS.UnitTest/Monitoring/Test_FileMonitor.cs]: Does this mean, that the Task, has startet the monitoringAsync?

File: RTDS/RTDS/Monitoring/Wrapper/TimerWrapper.cs [hunk 20]
```diff
@@ -0,0 +1,42 @@
...
+            _timer.Start();
+        }
+
+        public TimerWrapper(Timer timer)
+        {
+            _timer = timer;
+            _timer.Elapsed += OnElapsed;
+        }
+
+        private void OnElapsed(object source, ElapsedEventArgs e) => Elapsed?.Invoke(source, e);
```
Comment (MikkelHarbo) [inline RTDS/RTDS/Monitoring/Wrapper/TimerWrapper.cs]: I dont feel like the naming convension here of the "e" gives so much information?

File: RTDS/RTDS/Monitoring/Wrapper/TimerWrapper.cs [hunk 21]
```diff
@@ -20,23 +25,17 @@ public bool Enabled
             set => _timer.Enabled = value;
         }
 
-        public async Task Reset()
+        public void Reset()
```
Comment (MikkelHarbo) [inline RTDS/RTDS/Monitoring/Wrapper/TimerWrapper.cs]: Why did you change it?

File: RTDS/RTDS/Monitoring/IFileSystemWatcherWrapper.cs (hunk 12 above)
Comment (ChristianMorup) [reply RTDS/RTDS/Monitoring/IFileSystemWatcherWrapper.cs]: It is basically just wrapping the events from the FileSystemWatcher - making it testable.

File: RTDS/RTDS/Monitoring/FileSystemWatcherWrapper.cs (hunk 14 above)
Comment (ChristianMorup) [reply RTDS/RTDS/Monitoring/FileSystemWatcherWrapper.cs]: See https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/properties . It is how properties work.

File: RTDS/RTDS/Monitoring/FileSystemWatcherWrapper.cs (hunk 15 above)
Comment (ChristianMorup) [reply RTDS/RTDS/Monitoring/FileSystemWatcherWrapper.cs]: See https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/properties . It is how properties work.

File: RTDS/RTDS/Monitoring/FolderMonitor.cs (hunk 16 above)
Comment (ChristianMorup) [reply RTDS/RTDS/Monitoring/FolderMonitor.cs]: It's just for preventing modification once it has been initialised. The rule basically is that if can be readonly, then let it.

File: RTDS/RTDS.IntegrationTest/Monitoring/Test_FileSystemWatcherWrapper.cs (hunk 18 above)
Comment (ChristianMorup) [reply RTDS/RTDS.IntegrationTest/Monitoring/Test_FileSystemWatcherWrapper.cs]: Yup. Otherwise the folder will bloat with files.

Comment (ChristianMorup) [reply RTDS/RTDS.IntegrationTest/Monitoring/Test_FileSystemWatcherWrapper.cs]: And fail the second time the test is run :)

File: RTDS/RTDS.UnitTest/Monitoring/Test_FileMonitor.cs (hunk 19 above)
Comment (ChristianMorup) [reply RTDS/RTDS.UnitTest/Monitoring/Test_FileMonitor.cs]: Well - It means that the uut (unit under test) / filemonitor has started monitoring async. The task is just a tool in the framework for doing async work.

File: RTDS/RTDS/Monitoring/Wrapper/TimerWrapper.cs (hunk 20 above)
Comment (ChristianMorup) [reply RTDS/RTDS/Monitoring/Wrapper/TimerWrapper.cs]: True, I'll change it to args.

File: RTDS/RTDS/Monitoring/Wrapper/TimerWrapper.cs (hunk 21 above)
Comment (ChristianMorup) [reply RTDS/RTDS/Monitoring/Wrapper/TimerWrapper.cs]: Because it makes no sense to return a Task :)
